File: apache/htdocs/status/boks_info.pl
#!/usr/bin/perl
 
use Pg;
use CGI qw/:standard :html3/;
#use Time::Local;

print header(),start_html("Boks - info");


$apache_dir = '/usr/local/nav/navme/apache';
$navstart = "$apache_dir/vhtdocs/navstart.pl";
$navslutt = "$apache_dir/vhtdocs/navslutt.pl";


$sysName = param("sysName");

###########################################
# Kjører filen navstart, og skriver "print-linjene" til web
print `$navstart`;
###########################################

#unless ($dbh = DBI->connect("DBI:mysql:manage","nett","stotte"))
#{
#    print "<h4>Fikk ikke kontakt med databasen.</h4><p>";
#    print "<hr><i>grohi\@itea.ntnu.no</i>";
#    exit(0);
#}

$db = "manage";
$conn = db_connect($db);


$sql = "SELECT ip,boks.romid,typeid,orgid,kat,kat2,sted.descr,boksid,rom.descr FROM boks join rom USING (romid) join sted USING (stedid) WHERE sysName=\'$sysName\'";

$boksinfo = db_select($sql,$conn);

unless ($boksinfo->ntuples) # treff i databasen sysName
{
    print b("Ingen treff på $sysName."),p;
    # Avslutter $conn "old style":
    PQfinish($conn);
    print `$navslutt`;
    exit();

}
else
{
#    @rows = th({bgcolor=>e6e6e6},['navn','ip','nede siden','nedetid','info']);
    while (@svar = $boksinfo->fetchrow)
    {
	($ip,$rom,$type,$eier,$kat,$kat2,$sted,$boksid,$rombeskr)=@svar;
    }

    print "<h2><b>$sysName</b></h2><p>";

#print "$ip,$rom,$type,$eier,$kat,$kat2,$sted,$boksid<br>";

    @rows = ();
    
    push (@rows,td(["IP:",$ip]));
    push (@rows,td(["Type:",$type]));
    push (@rows,td(["Plassering:","$rom: $rombeskr, $sted"]));
    push (@rows,td(["Kategori:",$kat]));
    push (@rows,td(["Kategori2:",$kat2]));
    push (@rows,td(["Eier/Drifter:",uc($eier)]));
    


}
    print table({-border=>0 -cellpadding=>2 -cellspacing=>1},
	#    caption(b('Bokser nede')),
	    Tr(\@rows)
	    ),p;

#############
# Uplink - info

#$upsok = $dbh->prepare('select sysname,mp from swport,nettel where nettel.id=swid and portname like "n%" and idbak=?');
#$upsok->execute($nettelid);

#if ($upsok->rows)
#{
#    print "<tr><td>Uplink</td><td>";
#    while (@svar=$upsok->fetchrow_array)
#    {
#	print "<a href=/live/boks_info.pl?sysName=$svar[0]>$svar[0]</a> $svar[1]<br>";
#    }
#    print "</td></tr>";
#}
##############
#print "<tr><td colspan=2><a href=/live/hist_stat.pl?sysName=$sysName>Historikk</a></td></tr>";

#if ($kat eq 'HUB')
#{
#    &hent_org_kat;
#    ($org_enhet,$kat_enhet,$dummy) =split(/\-/,$sysName);
    
#    print "<tr><td valign=top>Funksjon:</td><td>Nettverkshub/svitsj for <b>$kats{$kat_enhet}</b> ved<br>";
#    print "<b>$orgs{$org_enhet}</b> som<br> sogner til telematikkrom i<br> $sted</td></tr>";
#}
#elsif ($kat eq 'SRV')
#{
#    print "<tr><td align=top>Funksjon:</td><td>$function</td></tr>";
#}

#print "</table>";

##############################
#if ($kat eq 'GW')
#{
#    &hent_org_kat;
#    print "<h3>Brukere bak $sysName:</h3>";
#
#    $gwsok = $dbh->prepare('SELECT vlan,gwip,org,kat,komm,bits FROM subnet WHERE type=? AND ruter=? order by vlan'); 
#    $gwsok->execute('lan',$nettelid);
#
#    print "<table>";
#    print "<tr><th>Brukere</th><th>vlan</th><th>gwip/maske</th></tr>";
#    while (@line=$gwsok->fetchrow_array)
#    {
#	if ($line[3] =~ /\d$/)
#	{
#	    chop($line[3]);
#	}
#	print "<tr><td><li>$kats{$line[3]} ved $orgs{$line[2]}";
#
#	print " ($line[4])" if $line[4];
#	print "</td>";
#
#	print "<td>$line[0]</td><td>$line[1]/$line[5]</td></tr>";



##	print "<tr><td>$line[0]</td><td>$line[1]</td><td>$line[2]</td><td>$line[3]</td><td>$line[4]</td></tr>";
#    }
#    print "</table>";
#}

###############################
#if ($kat eq 'SW')
#{
#    &hent_org_kat;
#
#    $swsok=$dbh->prepare('SELECT distinct sw.vlan,org,kat,komm FROM swport sw,subnet su WHERE sw.vlan=su.vlan AND su.type=? AND sw.vlan!=? AND status=? AND swid=? ORDER BY vlan');
#    $swsok->execute('lan','Trunk','Up',$nettelid);#
#
#    while (@line=$swsok->fetchrow_array)
#    {
#	$vlan{$line[0]}{org}  = $line[1];
#	$vlan{$line[0]}{kat}  = $line[2];
#	$vlan{$line[0]}{komm} = $line[3];
#    }
#    $swsok->finish;

#    $swsok2 = $dbh->prepare('SELECT distinct trunk.vlan,subnet.org,subnet.kat,subnet.komm FROM swport,nettel,trunk,subnet WHERE swport.vlan=\'Trunk\' AND swportid=swport.id AND swid=nettel.id AND subnet.vlan=trunk.vlan AND portname like "n:%" AND swport.status=\'Up\' AND sysName=? AND subnet.org IS NOT NULL ORDER BY trunk.vlan');

#    $swsok2->execute($sysName);

#    while (@line=$swsok2->fetchrow_array)
#    {
#	$vlan{$line[0]}{org}  = $line[1];
#	$vlan{$line[0]}{kat}  = $line[2];
#	$vlan{$line[0]}{komm} = $line[3];
#    }
#    $swsok2->finish;


#    print "<h3>Brukere bak $sysName:</h3>";
#    print "<table>";
#    print "<tr><th>Brukere</th><th>vlan</th></tr>";#
#
#    foreach $vlan (sort by_number keys %vlan)
#    {
#	print "<tr>";

#	if ($vlan{$vlan}{kat} =~ /\d$/)
#	{
#	    chop($vlan{$vlan}{kat});
#	}
#	print "<td><li>$kats{$vlan{$vlan}{kat}} ved $orgs{$vlan{$vlan}{org}}";

#	print " ($vlan{$vlan}{komm})" if $vlan{$vlan}{komm};
#	print "</td>";
#	print "<td>$vlan</td></tr>";
#    }
#    print "</table>";
    
#}


# Avslutter $conn "old style":
PQfinish($conn);
###########################################
# Kjører filen navslutt, og skriver "print-linjene" til web
print `$navslutt`;
###########################################

#print end_html;


##############################################
##############################################

sub hent_org_kat
{
    $orgsok=$dbh->prepare('SELECT org,descr FROM organisasjon');
    $orgsok->execute();
    while (@line=$orgsok->fetchrow_array)
    {
	$orgs{$line[0]} = $line[1];
    }
    $orgsok->finish;
    
    $katsok=$dbh->prepare('SELECT kat,descr FROM brukerkat');
    $katsok->execute();
    while (@line=$katsok->fetchrow_array)
    {
	$kats{$line[0]} = $line[1];
    }
    $katsok->finish;
}

sub by_number
{
    if ($a < $b)
    { return -1 }
    elsif ($a == $b)
    { return 0 }
    elsif ($a > $b)
    { return 1 }
}

##########################################################

sub db_connect {
    my $db = $_[0];
    my $conn = Pg::connectdb("dbname=$db user=navall password=uka97urgf");
    die $conn->errorMessage unless PGRES_CONNECTION_OK eq $conn->status;
    return $conn;
}
sub db_select {
    my $sql = $_[0];
    my $conn = $_[1];
    my $resultat = $conn->exec($sql);
    die "DATABASEFEIL: $sql\n".$conn->errorMessage
        unless ($resultat->resultStatus eq PGRES_TUPLES_OK);
    return $resultat;
}
sub db_execute {
    my $sql = $_[0];
    my $conn = $_[1];
    my $resultat = $conn->exec($sql);
    die "DATABASEFEIL: $sql\n".$conn->errorMessage
        unless ($resultat->resultStatus eq PGRES_COMMAND_OK);
    return $resultat;
}

File: apache/htdocs/status/boksinfo.php
<?php

require('/usr/local/nav/navme/apache/vhtdocs/nav.inc');

if (!$bruker) {
  $bruker = $PHP_AUTH_USER;
}

$vars = $HTTP_GET_VARS;

$sysName = $vars[sn];

if (!$sysName)
{
  navstart("Boksinfo",$bruker);
  print "SysName mangler!<br>";
}
else
{

  navstart("Boksinfo for $sysName",$bruker);
#  print "SysName: $sysName<br>";

  $dbh = pg_Connect ("dbname=manage user=navall password=uka97urgf");

  boksdata($dbh,$sysName);

}

?>


<?php

navslutt();

?>

<?php

function boksdata($dbh_,$sysName)
{
  
    $sql = "SELECT ip,typeid,boks.romid,rom.descr,sted.descr,kat,kat2,orgid,boksid FROM boks join rom USING (romid) join sted USING (stedid) WHERE sysName='$sysName'";


  $result = pg_exec($dbh_,$sql);
  $rows = pg_numrows($result);

    if ($rows == 0)
  {
    print "<b>Ingen treff på $sysName</b><br>";
  }  
  else
  {
     print "<h2><b>$sysName</b></h2><p>";

     print "<table border=0 cellpadding=2 cellspacing=1>";

     for ($i=0;$i < $rows; $i++) {
       $svar = pg_fetch_row($result,$i);

       $kat = $svar[5];

       print "<tr><td>IP:</td><td>$svar[0]</td></tr>";
       print "<tr><td>Type:</td><td>$svar[1]</td></tr>";
       print "<tr><td>Plassering:</td><td>$svar[2]: $svar[3], $svar[4]</td></tr>";
       print "<tr><td>Kategori:</td><td>$svar[5]</td></tr>";
       print "<tr><td>Kategori2:</td><td>$svar[6]</td></tr>";
       print "<tr><td>Eier/drifter:</td><td>$svar[7]</td></tr>";
     }
     print "</table>";

     print "<a href=./historikk.php?sn=$sysName>Historikk siste 7 dager</a>";

     print "<hr>";

     if ($kat == 'GW')
     {
       gwinfo($dbh_,$sysName,$svar[8]);
     }
     if ($kat == 'SW')
     {
       swinfo($dbh_,$sysName,$svar[8]);
     }


  }
}

#################################
#################################
#################################

function swinfo($dbh_,$sysName,$boksid)
{
  $sql = "SELECT distinct anv.descr,org.descr,vlan FROM swport JOIN swportvlan USING (swportid) JOIN prefiks USING (vlan) JOIN org USING (orgid) JOIN anv USING (anvid) WHERE status='up' AND boksid=$boksid";

  $result = pg_exec($dbh_,$sql);
  $rows = pg_numrows($result);

  if ($rows == 0)
  {
    print "<b>Ingen treff på brukere bak $sysName</b><br>";
  }  
  else
  {
    print "<b>brukere bak $sysName</b><p>";

    print "<table border=0 cellpadding=2 cellspacing=1>";
    print "<tr><th>brukergruppe</th><th>organisasjon</th><th>vlan</th></tr>";
    for ($i=0;$i < $rows; $i++) {
      $svar = pg_fetch_row($result,$i);
      print "<tr><td>$svar[0]</td><td align=center>$svar[1]</td><td align=right>$svar[2]</td></tr>";
    }
    print "</table>";
  }

}
#################################

function gwinfo($dbh_,$sysName,$boksid)
{

  $sql = "SELECT anv.descr,org.descr,nettadr,maske FROM gwport JOIN prefiks USING (prefiksid) JOIN anv USING (anvid) JOIN org USING (orgid) WHERE rootgwid=gwportid AND boksid=$boksid";

  $result = pg_exec($dbh_,$sql);
  $rows = pg_numrows($result);

  if ($rows == 0)
  {
    print "<b>Ingen treff på brukere/nett bak $sysName</b><br>";
  }  
  else
  {
    print "<b>brukere/nett bak $sysName</b><p>";

    print "<table border=0 cellpadding=2 cellspacing=1>";
    print "<tr><th>brukergruppe</th><th>organisasjon</th><th>nett/maske</th></tr>";
    for ($i=0;$i < $rows; $i++) {
      $svar = pg_fetch_row($result,$i);
      print "<tr><td>$svar[0]</td><td align=center>$svar[1]</td><td>$svar[2]/$svar[3]</td></tr>";
    }
    print "</table>";
  }
}

?>